UI Automation in Windows Console: limit blank lines in review and initial word movement support

codeofdusk opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
Split from #9646 (builds on #9614).

### Summary of the issue:
Currently, in consoles with UI Automation enabled:

* In object review, commands for moving to and reporting words move to or report lines instead.
* A large number (thousands in many cases) of empty lines are appended to the console's object review.

### Description of how this pull request fixes the issue:
Word movement and blank line filtering have been implemented in an overridden `move` method on `consoleUIATextInfo`. Since the console's UIA implementation does not provide word movement, we must use an offset-based word movement algorithm which uses the Uniscribe API to find word boundaries. The algorithm used for this implementation is based on that in `NVDAObjects.offsets._getWordOffsets`. Blank lines are filtered by comparing the `textInfo`'s `_rangeObj` to the last visible text range (returned by `winConsoleUIA.UIATextPattern.GetVisibleRanges`).

### Testing performed:
Tested text review on Windows 10 versions 1803 and 1903.

### Known issues with pull request:
* While "review previous word" and "review next word" now allow word movement, the cursor is not consistently placed on the first character of the word under review.

See #9646 for all console UIA known issues to date.
### Change log entry:
None.

codeofdusk committed:
Restrict the number of blank lines shown in console UIA's object review.

Note: there are issues when reaching the bottom of the review (cursor gets stuck).

codeofdusk committed:
UIA console: only apply our custom move logic when not moving the caret.

This should mostly restore caret movement support.

codeofdusk committed:
Add basic word movement support.

codeofdusk committed:
ConsoleUIATextInfo: support expantion by word.

codeofdusk committed:
consoleUIATextInfo._countCharsToEnd(reverse=True) is now consoleUIATextInfo._getCurrentOffset. The reverse option has been removed.

codeofdusk committed:
Cleanup.

codeofdusk commented:
CC @feerrenrut, @leonardder, @michaelDCurran.

codeofdusk committed:
Improve cursor placement after word movement.

michaelDCurran commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -30,6 +35,115 @@ def __init__(self, obj, position, _rangeObj=None):
+		return super(consoleUIATextInfo, self).move(unit, direction, endPoint)
+
+	def expand(self, unit):
+		if unit == textInfos.UNIT_WORD:
```
Expand should use similar code to move. E.g.:
If unit is word, then use _getCurrentOffset and _getWordOffsets and manually move the start and end of the range to have it correctly bound the current word.
Then, there is no need to change what text is exposed as the textRange will be correct.
The advantage of this is there is less to track, it is easier to read, and most importantly, later comparisons or movements of this textInfo won't behave like it is expanded to line even though the text returned is word.
There may be a question around performance: but I'd like to see the code changed to my proposed way before we then consider if optimisations are needed at all.

codeofdusk committed:
Review actions (without changes to expand)

codeofdusk committed:
Re-implement expand.

Note: this new approach includes characters after a line break in the current word.

michaelDCurran commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -140,8 +151,8 @@ def _getCurrentWord(self):
 		lineInfo = self.copy()
```
I think _getCurrentWord can be removed now as it is no longer used?

michaelDCurran commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -64,36 +61,50 @@ def move(self, unit, direction, endPoint=None):
+			self._rangeObj.MoveEndpointByUnit(
+				UIAHandler.TextPatternRangeEndpoint_Start,
+				UIAHandler.NVDAUnitsToUIAUnits[textInfos.UNIT_CHARACTER],
+				wordEndPoints[0]
```
What happens if wordEndPoints[0] or wordEndPoints[1] is 0? I.e. you don't ned to move? perhaps these moveByUnit calls should be within an if conndition checking that wordEndPoints[x] is not 0.

michaelDCurran commented:
when moving by word, the first two characters of the next line are included in the last word on a line.
The reason for this is to do with  the fact that an extra two alphanumeric characters are placed at the end of the text with give to calculateWordOffsets to get around a uniscribe bug. However, when returning the word offsets, len(lineText) is used. This should be len(lineText)-2 as we don't want to include the inserted characters.

michaelDCurran commented:
master will need to be merged into this branch, and any conflicts resolved.

codeofdusk committed:
Review actions.

codeofdusk committed:
Merge branch 'master' into cmduia2-textinfos

codeofdusk commented:
@michaelDCurran It's almost perfect now (I think the word movement bugs we were talking about earlier are fixed by the new implementation of `expand`). Only problem now is when you move past the last word on a line, NVDA repeatedly says "blank" (moving to the previous word puts us on the last line though).

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -17,10 +19,12 @@
 
 class consoleUIATextInfo(UIATextInfo):
 	_expandCollapseBeforeReview = False
+	_isCaret = False
```
I don't think this is necessary, see below.

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -31,6 +35,121 @@ def __init__(self, obj, position, _rangeObj=None):
 
+	def move(self, unit, direction, endPoint=None):
+		oldRange=None
+		if not self._isCaret:
```
Note that every TextInfo instance has an attribute `basePosition` that should be textInfos.POSITION_CARET. Therefore, I think I'd prefer this:

```suggestion
		if self.basePosition != textInfos.POSITION_CARET:
```

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -31,6 +35,121 @@ def __init__(self, obj, position, _rangeObj=None):
+			# Insure we haven't gone beyond the visible text.
+			# UIA adds thousands of blank lines to the end of the console.
+			visiRanges = self.obj.UIATextPattern.GetVisibleRanges()
+			if visiRanges.length > 0:
```
I don't know how caching works here, but it might make sense to do something like this:

```suggestion

			visibleRangesLength = visiRanges.Length
			if visibleRangesLength > 0:
```

codeofdusk replied:
What exactly is the advantage of doing this? We only use the value once per function call.

leonardder replied:
You're also getting the length for lastVisiRange

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -31,6 +35,121 @@ def __init__(self, obj, position, _rangeObj=None):
+			# UIA doesn't implement word movement, so we need to do it manually.
+			offset = self._getCurrentOffsetInThisLine()
+			start, end = self._getWordOffsetsInThisLine(offset)
+			wordEndPoints = (
```
It might make sense to save these as _wordEndPoints on the class, so you don't have to define them twice. Easy to overlook if it ever needs to be changed.

codeofdusk replied:
`wordEndPoints` and `wordMoveDirections` are slightly different (we subtract one from the former to make `expand` work).

leonardder replied:
Ah, thanks for clarifying.

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -31,6 +35,121 @@ def __init__(self, obj, position, _rangeObj=None):
+		charInfo = self.copy()
+		res = 0
+		chars = None
+		while True:
```
Recent experiences with Notepad++ made me a bit worried about endless while loops. Could you limit this somehow?

Would something like this work?

```suggestion
		while charInfo.compareEndPoints(
			lineInfo,
			"startToEnd"
		) <= 0:
```

leonardder commented on source/NVDAObjects/UIA/winConsoleUIA.py:
```diff
@@ -31,6 +35,121 @@ def __init__(self, obj, position, _rangeObj=None):
+		else:
+			return super(consoleUIATextInfo, self).expand(unit)
+
+	def _getCurrentOffsetInThisLine(self):
```
In most if not all cases, you're calling _getCurrentOffsetInThisLine and then _getWordOffsetsInThisLine, which both copy self and expand to a line. May be lineInfo could be an argument to both functions to avoid this? Of course, it should still only be used when self is collapsed within that line

codeofdusk committed:
Review actions.

codeofdusk committed:
Review actions.

codeofdusk committed:
Merge branch 'master' into cmduia2-textinfos

michaelDCurran reviewed (changes requested):
When moving forward by word and you land on a blank line, moving forward by word again gets stuck on the blank line. The reason for this is that although UIA textRange's treat a blank line as one character unit, the text it gives back for this unit is of 0 length. Therefore getWordOffsetsInThisLine ends up giving back 0,0.  
When fetching the line text in getWordOffsetInThisLine, set lineText to u" " if lineText is of 0 length. Easiest way is to do:
```
lineText=info.text or u" "
```

codeofdusk committed:
Fix forward movement across lines.

codeofdusk commented:
@michaelDCurran Thanks. Now, when moving forward by word, we can land (and pass over) the blank, but moving back by word skips over it. I think this is pretty minor though and this PR is ready for another review/merge.

michaelDCurran committed:
UIAWinConsole support: Improve logic of moving previous word to no longer jump over the last word on the previous line.

michaelDCurran commented:
I think I have come up with a fix for moving previous word that does not skip the last word of the previous line.
You can find the code in the pr9647_fixMovePreviousWord branch on nvaccess/nvda repository.
I have provided comments through out the code so you can hopefully understand the logic.
It was rather tricky to get right. Essentially it no lonver recursively calls move within the algorithm as in some cases that was happening too many times, plus the steps are now broken down into very small / easy to understand parts. 
In short the steps are now:
* Relative to the current line, calculate our offset and the current word's offsets.
* If moving in a forward direction, we can just jump to the end offset of the current word and we're done.
* However, if moving backwards:
 * If we are after the beginning of a word, first move back to the start of the word.
 * Try to move one character back before the start of the word.
 * If we have moved on to a previous line, recalculate our new offset for the new line.
 * Finally using the offset, Calculate the current word offsets and move to the start of this word if we are not already there.

michaelDCurran commented:
If you are happy with the code, and can follow the logic okay, please merge it into this pr and I will then approve and merge this pr to master.

michaelDCurran commented:
This would have been a lot easier to write just using expand and collapse calls within move, but Microsoft's UIA implementation for consoles seems to not handle collapsing properly, as you find with the review cursor scripts previously.

codeofdusk committed:
Merge remote-tracking branch 'origin/pr9647_fixMovePreviousWord' into cmduia2-textinfos

codeofdusk committed:
Style.

codeofdusk commented:
@michaelDCurran Looks good to me!
